snapshots/btrfs/btrfs.go

```diff
@@ -345,65 +319,59 @@ func (b *snapshotter) Remove(ctx context.Context, key string) (err error) {
 		readonly        bool
 	)
 
-	ctx, t, err := b.ms.TransactionContext(ctx, true)
-	if err != nil {
-		return err
-	}
 	defer func() {
-		if err != nil && t != nil {
-			if rerr := t.Rollback(); rerr != nil {
-				log.G(ctx).WithError(rerr).Warn("failed to rollback transaction")
-			}
-		}
-
 		if removed != "" {
 			if derr := btrfs.SubvolDelete(removed); derr != nil {
 				log.G(ctx).WithError(derr).WithField("subvolume", removed).Warn("failed to delete subvolume")
 			}
 		}
 	}()
 
-	id, k, err := storage.Remove(ctx, key)
-	if err != nil {
-		return fmt.Errorf("failed to remove snapshot: %w", err)
-	}
+	err = b.ms.WithTransaction(ctx, true, func(ctx context.Context) error {
+		id, k, err := storage.Remove(ctx, key)
+		if err != nil {
+			return fmt.Errorf("failed to remove snapshot: %w", err)
+		}
 
-	switch k {
-	case snapshots.KindView:
-		source = filepath.Join(b.root, "view", id)
-		removed = filepath.Join(b.root, "view", "rm-"+id)
-		readonly = true
-	case snapshots.KindActive:
-		source = filepath.Join(b.root, "active", id)
-		removed = filepath.Join(b.root, "active", "rm-"+id)
-	case snapshots.KindCommitted:
-		source = filepath.Join(b.root, "snapshots", id)
-		removed = filepath.Join(b.root, "snapshots", "rm-"+id)
-		readonly = true
-	}
+		switch k {
+		case snapshots.KindView:
+			source = filepath.Join(b.root, "view", id)
+			removed = filepath.Join(b.root, "view", "rm-"+id)
+			readonly = true
+		case snapshots.KindActive:
+			source = filepath.Join(b.root, "active", id)
+			removed = filepath.Join(b.root, "active", "rm-"+id)
+		case snapshots.KindCommitted:
+			source = filepath.Join(b.root, "snapshots", id)
+			removed = filepath.Join(b.root, "snapshots", "rm-"+id)
+			readonly = true
+		}
 
-	if err := btrfs.SubvolSnapshot(removed, source, readonly); err != nil {
-		removed = ""
-		return err
-	}
+		if err = btrfs.SubvolSnapshot(removed, source, readonly); err != nil {
+			removed = ""
+			return err
+		}
 
-	if err := btrfs.SubvolDelete(source); err != nil {
-		return fmt.Errorf("failed to remove snapshot %v: %w", source, err)
-	}
+		if err = btrfs.SubvolDelete(source); err != nil {
+			return fmt.Errorf("failed to remove snapshot %v: %w", source, err)
+		}
+		return nil
```

Comment:
sorry for missing this one.

At the original L397, the restore source only happens when it fails to commit change into boltdb.
I think we should use `var restore bool` and set it to `true`. And then 

```
if err != nil {
     if restore  {
            // Attempt to restore source
            if err1 := btrfs.SubvolSnapshot(source, removed, readonly); err1 != nil {
            }
     }
}
``` 

It can align with original behaviour

Reply:
New changes have been pushed. I don't know if it matches your review.